app/controllers/users/phone_confirmation_controller.rb

```diff
@@ -0,0 +1,97 @@
+module Users
+  class PhoneConfirmationController < ApplicationController
+    before_action :authenticate_user!
+    before_action :check_for_unconfirmed_mobile
+
+    def show
+      if FeatureManagement.prefill_otp_codes?
+        @code_value = confirmation_code
+      end
+
+      @unconfirmed_mobile = unconfirmed_mobile
+      @reenter_phone_number_path = if current_user.mobile.present?
+                                     edit_user_registration_path
+                                   else
+                                     users_otp_path
+                                   end
+    end
+
+    def new
+      send_new_confirmation_code
+      redirect_to phone_confirmation_path
+    end
+
+    def confirm
+      if params['code'] == confirmation_code
+        process_valid_code
+      else
+        process_invalid_code
+      end
+    end
+
+    def self.generate_confirmation_code
+      digits = Devise.direct_otp_length
+      random_base10(digits)
+    end
+
+    def self.random_base10(digits)
+      SecureRandom.random_number(10**digits).to_s.rjust(digits, '0')
+    end
+
+    private
+
+    def process_invalid_code
+      flash[:error] = t('errors.messages.invalid_confirmation_code')
+      redirect_to phone_confirmation_path
+    end
+
+    def process_valid_code
+      flash[:success] = t('upaya.notices.phone_confirmation_successful')
+      @updating_existing_number = current_user.mobile.present?
+      current_user.mobile_confirm(unconfirmed_mobile)
+      clear_session_data
+      redirect_to after_confirmation_path
+    end
+
+    def after_confirmation_path
+      if @updating_existing_number
+        edit_user_registration_path
+      else
+        after_sign_in_path_for(current_user)
+      end
+    end
+
+    def check_for_unconfirmed_mobile
+      redirect_to root_path unless unconfirmed_mobile
```

Comment:
This cannot happen under normal circumstances so i don't think it matters, and they won't be able to get to after_sign_in_path_for(user) if they are not fully signed in anyway.